src/Native/Unix/Common/pal_safecrt.h

```diff
@@ -15,7 +16,7 @@
 // Returns true if safe, false if overflows.
 inline static bool multiply_s(size_t a, size_t b, size_t* result)
 {
-#if __has_builtin(__builtin_mul_overflow)
+#if HAVE_BUILTIN_MUL_OVERFLOW
```

Comment:
This check is failing in CI in the Linux x64 leg:
```
-- Performing Test HAVE_BUILTIN_MUL_OVERFLOW - Failed
```
Wouldn't we expect it to succeed there?

Reply:
This seems to aligned with the master branch. The Linux x64 leg is using Clang 3.6.2 (release Jul 2015) on ` microsoft/dotnet-buildtools-prereqs:rhel7_prereqs_2` container, that runs Red Hat Enterprise Linux 7.2 (release Nov 2015).

A test with master:

```sh
docker run -it microsoft/dotnet-buildtools-prereqs:rhel7_prereqs_2

#inside docker
git clone https://github.com/dotnet/corefx --single-branch --depth 1 --branch master
cd corefx

#apply a patch
git apply <<EOF
diff --git a/src/Native/Unix/Common/pal_safecrt.h b/src/Native/Unix/Common/pal_safecrt.h
index f0627e4..67a19d7 100644
--- a/src/Native/Unix/Common/pal_safecrt.h
+++ b/src/Native/Unix/Common/pal_safecrt.h
@@ -16,8 +16,10 @@
 inline static bool multiply_s(size_t a, size_t b, size_t* result)
 {
 #if __has_builtin(__builtin_mul_overflow)
+#error "has __builtin_mul_overflow"
     return !__builtin_mul_overflow(a, b, result);
 #else
+#error "no __builtin_mul_overflow (slow path)"
     if (a == 0 || b == 0)
     {
         *result = 0;
EOF

src/Native/build-native.sh x64 checked
```

it hits the `#else` case:

```sh
[ 40%] Building C object System.Native/CMakeFiles/System.Native.dir/pal_interfaceaddresses.c.o
In file included from /corefx/src/Native/Unix/System.Native/pal_interfaceaddresses.c:9:
/corefx/src/Native/Unix/Common/pal_safecrt.h:22:2: error: "no __builtin_mul_overflow (slow path)"
#error "no __builtin_mul_overflow (slow path)"
 ^
1 error generated.
```

Reply:
The reason for cmake autoconfig was that GCC does not provide `__has_builtin` macro, until today: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66970.

Reply:
The reason for cmake autoconfig was that GCC does not provide `__has_builtin` macro, even in the latest version as of today: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66970.

Reply:
I am quite surprised that we don't use the CentOS 7 container where we have clang 3.9 built from sources. That was intended to be used to build our product. We use it for coreclr and I have assumed we use it for corefx and core-setup too.
@echesakovMSFT do you know why we differ here?

Reply:
@janvorli No, unfortunately, I don't know

Reply:
Would it be feasible to add a separate job for gcc after this PR goes in? :)
In the current matrix of CoreFX job, (unsurprisingly) the slimmest container is Alpine 3.6, which already has gcc v6x installed implicitly; aports, gcc package is a dependency of CMake, which we already have.

Reply:
Maintaining more legs in PRs, etc. is not free. We want to make sure that we understand the reason why we are doing this.

Could you please help us understand why you care about being able to build with gcc?

Reply:
Portability - llvm toolchain is not readily available everywhere when compared to gcc. Case in point, [SmartOS](https://www.joyent.com/smartos) and other Solaris-like Unices. After a (years) long wait, recently, llvm 6 was ported to pkgsrc, targeting SmartOS. That port still lacks some tools like lldb (which apparently requires fixes in lldb sources to support Solaris-likes).

[perhaps a bit subjective argument, but] From the perspective of produced binaries, speed-wise, the recent versions of gcc do outperform clang of same era in certain microbenchmarks: https://www.phoronix.com/scan.php?page=article&item=gcc-clang-eoy2017 & https://www.phoronix.com/scan.php?page=article&item=gcc9-clang8-aarch64.
(at the minimum, we would have a variant of same binary to compare with periodically)

Reply:
> Portability - llvm toolchain is not readily available everywhere when compared to gcc. Case in point, SmartOS and other Solaris-like Unices

It would require quite a bit of work to make .NET Core work on theses OSes. Our bar for adding CI coverage for new OSes has been that the platform needs to be somewhat working (e.g. HelloWorld runs).